dotcom-rendering/scripts/islands/descriptions.mjs

```diff
@@ -0,0 +1,146 @@
+import { fileURLToPath } from 'node:url';
+import rehypeAutolinkHeadings from 'rehype-autolink-headings';
+import rehypeSlug from 'rehype-slug';
+import rehypeStringify from 'rehype-stringify';
+import remarkParse from 'remark-parse';
+import remarkRehype from 'remark-rehype';
+import { unified } from 'unified';
+import * as zod from 'zod';
+
+/** @type {(name: string) => RegExp} */
+export const get_JSDoc = (name) =>
+	new RegExp(
+		'\n/\\*\\*\n?' +
+			'((?: \\*.*\n)+)' +
+			' \\*\\/' +
+			`\nexport const ${name} =`,
+		'm',
+	);
+
+const components_directory = resolve(
+	fileURLToPath(import.meta.url),
+	'../../../src/web/components',
+);
+
+const PREFIX = 'window.chartData = ';
+
+const chunk = zod.object({
+	label: zod.string(),
+	isAsset: zod.boolean(),
+	statSize: zod.number(),
+	parsedSize: zod.number(),
+	gzipSize: zod.number(),
+});
+
+const bundle_report = await readFile(
+	resolve(
+		fileURLToPath(import.meta.url),
+		'../../../dist/browser.modern-bundles.html',
+	),
+	'utf-8',
+)
+	.then((file) =>
+		file
+			.split('\n')
+			.find((line) => line.trim().startsWith(PREFIX))
+			?.trim(),
+	)
+	.then((bundle_data) =>
+		zod
+			.array(chunk)
+			.parse(
+				JSON.parse(
+					bundle_data?.slice(PREFIX.length, -1).trim() ?? '[]',
+				),
+			),
+	);
+
+const processor = unified()
+	.use(remarkParse)
+	.use(remarkRehype)
+	.use(rehypeSlug)
+	.use(rehypeAutolinkHeadings, { behavior: 'wrap' })
+	.use(rehypeStringify);
+
+const islands = await readdir(components_directory)
+	.then((components) =>
+		components.filter((component) => component.endsWith('.importable.tsx')),
+	)
+	.then((files) =>
+		Promise.all(
+			files.map(async (file) => ({
+				file,
+				content: await readFile(
+					resolve(components_directory, file),
+					'utf-8',
+				),
+			})),
+		),
+	)
+	.then((files_and_content) =>
+		Promise.all(
+			files_and_content.map(async ({ file, content }) => {
+				const name = file.replace('.importable.tsx', '');
+				const { gzipSize = 0, parsedSize = 0 } =
+					bundle_report.find(({ label }) =>
+						label.startsWith(name + '-importable.'),
+					) ?? {};
+
+				const REGEX = get_JSDoc(name);
+
+				const matched = content.match(REGEX);
+
+				const { index, match } = matched
+					? { match: matched[1], index: matched.index ?? -1 }
+					: { match: undefined, index: -1 };
+
+				const lines = match
+					? '#' +
+					  [
+							content.slice(0, index).split('\n').length + 1,
+							content.slice(0, index + match.length).split('\n')
+								.length + 2,
+					  ]
+							.map((line_count) => `L${line_count}`)
+							.join('-')
+					: '';
+
+				const description =
+					matched?.[1]
+						?.split('\n')
+						.map((jsdoc_line) => jsdoc_line.slice(3))
+						.join('\n') ?? `# ${name} \n No description yet… 😢`;
+
+				const html = await processor.process(description);
+
+				return `<li>
+    <header>
+        <h4>${(gzipSize / 1024).toFixed(1)}kB gzip</h4>
+        <h4>${(parsedSize / 1024).toFixed(1)}kB parsed</h4>
```

Comment:
This might be for a future PR - but do we have a traffic light system here? As an engineer I want to know what is good or bad. 

This could be as a fixed budged e.g. 1kb and compared to other islands e.g. "Your island is 1.2kb smaller than the average, amazing!"

Reply:
I’ve picked three inflexion points, which roughly represent this:

- 🟢 good
- 1kB
- 🟠 needs improvement
- 5kB 
- 🔴 poor